src/BarrelCalorimeterInterlayers_geo.cpp

```diff
@@ -192,71 +216,65 @@ void buildFibers(Detector& desc, SensitiveDetector& sens, Volume& s_vol, xml_com
+  if (x_fiber.isSensitive()) {
+    f_vol_core.setSensitiveDetector(sens);
+  }
+  f_vol_core.setAttributes(desc, x_fiber.regionStr(), x_fiber.limitsStr(), x_fiber.visStr());
 
+
+  // Calculate number of divisions
+  auto        grid_div = getNdivisions(s_trd_x1, s_thick, 2.0 * cm, 2.0 * cm);
+  // Calculate polygonal grid coordinates (vertices)
+  auto        grids = gridPoints(grid_div.first, grid_div.second, s_trd_x1, s_thick, hphi);
   vector<int> f_id_count(grid_div.first * grid_div.second, 0);
   auto        f_pos = fiberPositions(f_radius, f_spacing_x, f_spacing_z, s_trd_x1, s_thick, hphi);
-  // std::cout << f_pos.size() << " lines, ~" << f_pos.front().size() << " fibers each line" << std::endl;
-  for (size_t il = 0; il < f_pos.size(); ++il) {
-    auto& line = f_pos[il];
-    if (line.empty()) {
-      continue;
-    }
-    double l_pos_y = line.front().y();
-    // use assembly as intermediate volume container to reduce number of daughter volumes
-    Assembly lfibers_clad(Form("fiber_clad_array_line_%lu", il));
-    Assembly lfibers_core(Form("fiber_core_array_line_%lu", il));
-    for (auto& p : line) {
-      int f_grid_id = -1;
-      int f_id      = -1;
-      // Check to which grid fiber belongs to
-      for (auto& poly_vtx : grid_vtx) {
-        if (p.y() != l_pos_y) {
-          std::cerr << Form("Expected the same y position from a same line: %.2f, but got %.2f", l_pos_y, p.y())
-                    << std::endl;
-          continue;
-        }
-        auto [grid_id, vtx_a, vtx_b, vtx_c, vtx_d] = poly_vtx;
-        double poly_x[4]                           = {vtx_a.x(), vtx_b.x(), vtx_c.x(), vtx_d.x()};
-        double poly_y[4]                           = {vtx_a.y(), vtx_b.y(), vtx_c.y(), vtx_d.y()};
-        double f_xy[2]                             = {p.x(), p.y()};
-
-        TGeoPolygon poly(4);
-        poly.SetXY(poly_x, poly_y);
-        poly.FinishPolygon();
-
-        if (poly.Contains(f_xy)) {
-          f_grid_id = grid_id;
-          f_id      = f_id_count[grid_id];
-          f_id_count[grid_id]++;
-        }
+  // a helper struct to speed up searching
+  struct Fiber {
+    Point pos;
+    bool assigned = false;
+    Fiber (const Point &p) : pos(p) {};
+  };
+  std::vector<Fiber> fibers(f_pos.begin(), f_pos.end());
+
+  // build assembly for each grid and put fibers in
+  for (auto &gr : grids) {
+    Assembly grid_vol(Form("fiber_grid_%i_%i", gr.ix, gr.iz));
+    // fiber is along y-axis of the layer volume, so grids are arranged on X-Z plane
+    Transform3D gr_tr(RotationZYX(0, 0, M_PI * 0.5), Position(gr.mean_centroid.x(), 0, gr.mean_centroid.y()));
+    auto grid_phv = s_vol.placeVolume(grid_vol, gr_tr);
+    grid_phv.addPhysVolID(f_id_grid, gr.ix + gr.iz * grid_div.first + 1);
+
+    // loop over all fibers that are not assigned to a grid
+    int f_id = 1;
+    for (auto &fi : fibers | boost::adaptors::filtered( [](const Fiber &f) {return not f.assigned;} )) {
```

Comment:
No c++20 features for now. We can't use c++20 until ROOT can be installed with c++20 support.

Reply:
OK, I'll change these